### ISSUE TITLE
Annotation tap gesture recognizer queries hard-coded region, ignoring image sizes

### ISSUE THREAD

**Issue by jfirebaugh** [mapbox/mapbox-gl-native#1504]
Tapping on a point annotation in the iOS test app seems to open a callout as expected in only 15% or so of taps.

**Comment by incanus** [comment]
Can you get more specific about how close nearby annotations are visually, what point you are tapping (as logged in debug) and how that compares to the annotation you were trying to tap? 

Because annotations are in GL and aren't actual tappable views, we are querying a fuzzy region around the user's tap to try to guess which annotations they are referring to, as well as iterating overlapping annotations like MapKit does to enable selection of obscured markers. 

https://github.com/mapbox/mapbox-gl-native/blob/44cb91c656ecfa3a33340f837460b3f527874380/platform/ios/MGLMapView.mm#L993-L1117

In addition, this heuristic can vary depending on marker image size, as that affects the user's perceived tappable area while not changing the region query (though we [do weight the region](https://github.com/mapbox/mapbox-gl-native/blob/44cb91c656ecfa3a33340f837460b3f527874380/platform/ios/MGLMapView.mm#L1016-L1021) to account for top-heavy markers). 

In short, this has felt good for months now, so I'm curious what changed and what the specific scenario is that is triggering this impression.

**Comment by incanus** [comment]
More background: https://github.com/mapbox/mapbox-gl-native/pull/1037#issuecomment-84521846

**Comment by jfirebaugh** [comment]
I'm testing on device using the Emerald style and "Add 100 Points", and zooming into an area where there is only one marker. It looks like the problem is that the hitbox for the marker is about half the size of the actual marker in Emerald, so only taps on the lower half of the marker image register.

**Comment by incanus** [comment]
This is probably a result then of: 

1. Insufficient tap-testing with Emerald
1. Emerald's marker being so much bigger than Streets' (and Light/Dark's)

Streets: 

![](https://raw.githubusercontent.com/mapbox/mapbox-gl-styles/mb-pages/sprites/mapbox-streets%402x.png)

Emerald: 

![](https://raw.githubusercontent.com/mapbox/mapbox-gl-styles/mb-pages/sprites/emerald%402x.png)

However, this highlights the difficult with the stopgap method of limiting marker imagery to the sprite sheet while simultaneously trying to minimize communications between the GL sprite parsing layer and the iOS code relation to gestures. 

Possible solutions: 

* Tweak annotation gesture algorithms
* Come up with communications layer to better relate sprite sizing, but we will throw this away in b2 when we support runtime imagery (or will we?)
* Punt for b1, for Emerald
* Size down the Emerald marker imagery and document the limitation for users wanting their own sprite-based imagery /cc @peterqliu

**Comment by peterqliu** [comment]
I'm happy to size down the marker-- do we have a threshold in mind, both for the resizing and for the documentation?

**Comment by 1ec5** [comment]
We’ve got a [hardcoded 40×60 “tolerance”](https://github.com/mapbox/mapbox-gl-native/blob/master/platform/ios/MGLMapView.mm#L1013-1014) for tap recognition. It looks like the sprite is a bit taller than that.

**Comment by jfirebaugh** [comment]
I'm ok with punting on this for b1 for Emerald.

**Comment by incanus** [comment]
> I'm ok with punting on this for b1 for Emerald.

I think this is an ok course of action. So: no further actions here for b1, assigning issue to b2.

**Comment by incanus** [comment]
Again, b3 will feature point annotation improvements.

**Comment by 1ec5** [comment]
The situation may’ve improved a little at lower zoom levels due to #1827. `TransformState::latLngForPixel()` was being used by `-[MGLMapMap handleSingleTapGesture:]` to compute a bounds (centered around the tap) to pass into `Map::getAnnotationsInBounds()`.

**Comment by friedbunny** [comment]
This blocks large/oversized markers, such as this:

![screen shot 2015-07-10 at 1 14 11 pm](https://cloud.githubusercontent.com/assets/1198851/8625263/a7451616-270b-11e5-8dfb-4c4725724b5d.png)

Red square being an approximation of the allowed tap box.

**Comment by 1ec5** [comment]
MGLMapView:
* Hardcodes the slop area around the touch point to use as a bounding box
* Knows the dimensions of every custom marker image, but not the default marker sprite

mbgl::AnnotationManager:
* Considers shapes to have area but does not consider point annotations to have area (it should)

mbgl::SpriteAtlas:
* Knows the dimensions of the default marker sprite

**Comment by incanus** [comment]
I think we have all the parts we need here, save a possible radius-based query for annotation proximity (assisted by https://github.com/mapbox/mapbox-gl-native/pull/1694, I'm sure), then we can: 

1. Register a tap gesture. 
2. Query a suitable over-large radius for annotation candidates. 
3. Iterate those candidates' real regions to determine final candidates. 

This is relatively slow, but shouldn't be noticeable for reasonable annotation counts with regard to tap delays. I think?

**Comment by ljbade** [comment]
@incanus Now we have a fixed size default marker, as well as being able to directly know the size of custom markers from the bitmap, this should be implementable?

I'm going to take a stab on Android.

**Comment by ljbade** [comment]
@incanus I realised the hit box on Android had not been updated to match the size of the new default marker. Did this get updated on iOS?

**Comment by ljbade** [comment]
For Android https://github.com/mapbox/mapbox-gl-native/pull/2876 is the first part, it at least adjusts current hit box to be default marker + 10 dp on each side.

Next will be to change `getMarkersInBounds` to use the Bitmap size of each marker in mMarkers to determine if they are in the hit box.

**Comment by 1ec5** [comment]
Annotation images frequently have shadows or other ornamentation that shouldn’t be considered part of the hit target. The most natural way to handle this detail on iOS would be to set the image’s alignment rectangle; `-[MGLMapView handleSingleTapGesture:]` would then filter the results from `getPointAnnotationsInBounds()` by whether a given annotation’s image’s alignment rect intersects with the tap.

**Comment by 1ec5** [comment]
The OS X port in #3135 has a fix for this issue, but porting the fix back to iOS is blocked by #3159.

**Comment by 1ec5** [comment]
This will be fixed in #3261.
